src/postformatting/embedded-base64-remover.js

```diff
@@ -1,60 +1,61 @@
+import csstree from 'css-tree'
 import debug from 'debug'
 
 const debuglog = debug('penthouse:preformatting:embeddedbase64Remover')
 
-const BASE64_ENCODE_PATTERN = /data:[^,]*base64,/
-
-const _isTooLongBase64Encoded = function (declaration, maxEmbeddedBase64Length) {
-  const value = declaration.value.value
-  const tooLong =
-    BASE64_ENCODE_PATTERN.test(value) && value.length > maxEmbeddedBase64Length
-  if (tooLong) {
-    debuglog(
-      'DROP: ' +
-        `${declaration.property}: ${declaration.value.value.slice(0, 50)}..., (${value.length} chars)`
-    )
-  }
-  return tooLong
+const BASE64_ENCODE_PATTERN = /data:[^,]*;base64,/
+
+function hasSrc (node) {
+  return (
+    node.type === 'Declaration' &&
+    csstree.property(node.property).name === 'src'
+  )
 }
 
-const _removeDataUrisFromRule = function (rule, maxEmbeddedBase64Length) {
-  if (rule.type === 'Atrule' && rule.name === 'font-face') {
-    let hasSrc = false
-    rule.block.children = rule.block.children.filter(declaration => {
-      if (_isTooLongBase64Encoded(declaration, maxEmbeddedBase64Length)) {
-        return false
-      } else if (declaration.property === 'src') {
-        hasSrc = true
+const embeddedbase64Remover = function (ast, maxEmbeddedBase64Length) {
+  debuglog('config: maxEmbeddedBase64Length = ' + maxEmbeddedBase64Length)
+  csstree.walk(ast, {
+    visit: 'Declaration',
+    enter: (declaration, item, list) => {
+      let tooLong = false
+
+      csstree.walk(declaration, {
+        visit: 'Url',
+        enter: function (url) {
+          const value = url.value.value
+          if (
+            BASE64_ENCODE_PATTERN.test(value) &&
+            value.length > maxEmbeddedBase64Length
+          ) {
+            tooLong = true
+          }
+        }
+      })
+
+      if (tooLong) {
+        const value = csstree.generate(declaration.value)
+        debuglog(
+          'DROP: ' +
+            `${declaration.property}: ${value.slice(0, 50)}..., (${value.length} chars)`
+        )
+        list.remove(item)
       }
-      return true
-    })
-    if (!hasSrc) {
-      return null
     }
-  } else if (rule.type === 'Rule') {
-    rule.block.children = rule.block.children.filter(declaration => {
-      if (_isTooLongBase64Encoded(declaration, maxEmbeddedBase64Length)) {
-        return false
-      }
-      return true
-    })
-  } else if (rule.type === 'Atrule' && rule.name === 'media') {
-    rule.block.children = rule.block.children
-      .map(function (rule) {
-        return _removeDataUrisFromRule(rule, maxEmbeddedBase64Length)
-      })
-      .filter(Boolean)
+  })
 
-    return rule
-  }
-  return rule
-}
+  // remove @font-face atrules with no src declaration
+  csstree.walk(ast, {
```

Comment:
unnecessary to do another full `walk` on the whole ast for this - do it inside the `unused fontface remover` instead, I made a comment about it. Cheap since you area already iterating over the font-face rules.

Reply:
Agree, it can be moved it to font-face remover. But we need swap the order of ffRemover and embeddedbase64Remover, and change related tests, since embeddedbase64Remover tests check at-rules with no src is removed. If that's ok, I'll change it.

Reply:
Go for it 👍